lib/ecto/type.ex

```diff
@@ -10,55 +10,65 @@ defmodule Ecto.Type do
 
   ## Example
 
-  Imagine you want to support your id field to be looked up as a
-  permalink. For example, you want the following query to work:
+  Imagine you want to store an URI struct as part of a schema in an 
+  url-shortening service. There isn't an Ecto field type to support 
+  that value at runtime, therefore a custom one is needed.
 
-      permalink = "10-how-to-be-productive-with-elixir"
-      from p in Post, where: p.id == ^permalink
+  You also want to query not only by the full url, but for example 
+  by specific ports used. This is possible by putting the URI data
+  into a map field instead of just storing the plain 
+  string representation.
 
-  If `id` is an integer field, Ecto will fail in the query above
-  because it cannot cast the string to an integer. By using a
-  custom type, we can provide special casting behaviour while
-  still keeping the underlying Ecto type the same:
+      from s in ShortUrl,
+        where: fragment("?->>? ILIKE ?", s.original_url, "port", "443")
 
-      defmodule Permalink do
+  So the custom type does need to handle the conversion from 
+  external data to runtime data ([`cast/1`](#c:cast/1)) as well as 
```

Comment:
You can write the callbacks as `c:cast/1` and so on and ExDoc will be able to link to it.